### PR TITLE
Feature/pickup tests

### PR DESCRIPTION
**Pull request by jonbash**
## What's here (& why)

[Trello card](https://trello.com/c/CqTsTYlN/22-unit-tests-for-model-controller-for-pickup-scheduling)

- Further pickup tests
- Minor refactoring to support testing

### CODE REVIEW

File: EcoSoapBank.xcodeproj/xcshareddata/xcschemes/labs-ios-starter.xcscheme [hunk 1]
```diff
@@ -26,7 +26,8 @@
       buildConfiguration = "Debug"
       selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
       selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
-      shouldUseLaunchSchemeArgsEnv = "YES">
+      shouldUseLaunchSchemeArgsEnv = "YES"
+      codeCoverageEnabled = "YES">
```
Comment (swift-student) [inline EcoSoapBank.xcodeproj/xcshareddata/xcschemes/labs-ios-starter.xcscheme:30]: oooh, checking code coverage... 😃

File: EcoSoapBank/Model Controller/MockPickupProvider.swift [hunk 2]
```diff
@@ -57,6 +57,22 @@ extension Array where Element == Pickup {
...
+}
+
+extension Array where Element == Pickup.Carton {
+    static func random() -> [Pickup.Carton] {
+        [Pickup.CartonContents].random().map {
+            Pickup.Carton(id: UUID(), contents: $0)
+        }
+    }
+}
+
```
Comment (swift-student) [inline EcoSoapBank/Model Controller/MockPickupProvider.swift:75]: Nice extensions to give randomness for testing

File: EcoSoapBankTests/PickupTests.swift [hunk 3]
```diff
@@ -7,19 +7,97 @@
...
 
 class PickupTests: XCTestCase {
+    private var pickupCoordinator: PickupCoordinator!
+    private var pickupController: PickupController!
+    private var pickupProvider: MockPickupProvider!
+    private var mockDataExpectation: XCTestExpectation!
+    private var waiter: XCTWaiter!
+
+    override func setUp() {
+        super.setUp()
```
Comment (swift-student) [inline EcoSoapBankTests/PickupTests.swift:22]: Is this out of habit or is there a need to call super here?

File: EcoSoapBankTests/PickupTests.swift [hunk 4]
```diff
@@ -7,19 +7,97 @@
...
+
     func testPickupControllerSuccess() throws {
-        let pickupController = PickupController(dataProvider: MockPickupProvider())
-        XCTAssertEqual(pickupController.pickups.count, 3)
+        pickupProvider.fetchAllPickups { result in
+            self.mockDataReturned()
+
+            switch result {
+            case .failure(let error):
+                XCTFail("Mock pickup provider should have succeeded but failed with error: \(error)")
```
Comment (swift-student) [inline EcoSoapBankTests/PickupTests.swift:38]: Nice and descriptive

File: EcoSoapBankTests/PickupTests.swift [hunk 5]
```diff
@@ -7,19 +7,97 @@
...
+
+extension PickupTests {
+    private func waitForMockData() {
+        waiter.wait(for: [mockDataExpectation], timeout: 5)
+    }
+
+    private func mockDataReturned() {
+        mockDataExpectation.fulfill()
     }
 }
```
Comment (swift-student) [inline EcoSoapBankTests/PickupTests.swift:99]: Interesting setup. Can't say I would have thought of doing this.

File: EcoSoapBankTests/UtilityTests.swift [hunk 6]
```diff
@@ -28,4 +28,16 @@ class UtilityTests: XCTestCase {
         }
         XCTAssertEqual(object.int, 42)
     }
+
+    func testDate() {
```
Comment (swift-student) [inline EcoSoapBankTests/UtilityTests.swift:32]: Maybe rename this to be a bit more specific. I had to look in UtilityFunctions to realize that it was a custom init that you were testing.

``` swift
func testCustomDateInitializer()
```

File: EcoSoapBankTests/PickupTests.swift (hunk 3 above)
Comment (jonbash) [reply EcoSoapBankTests/PickupTests.swift:22]: SwiftLint yelled at me. 😅

File: EcoSoapBankTests/UtilityTests.swift (hunk 6 above)
Comment (jonbash) [reply EcoSoapBankTests/UtilityTests.swift]: Good call. Done.

**Pull request merged**